File: 힙/프로그래머스_이중 우선순위큐.py
from heapq import heappop, heappush, heapify

def solution(operations):
    Q = []
    for op in operations:
        m, num = op.split()
        num = int(num)

        if m == 'I':
            heappush(Q, num)

        else:
            if num < 0:
                if Q:
                    heappop(Q)
            else:
                if Q:
                    Q = list(map(lambda x: -x, Q))
                    heapify(Q)
                    heappop(Q)
                    Q = list(map(lambda x: -x, Q))
                    heapify(Q)


    if Q:
        return [max(Q), min(Q)]
    else:
        return [0,0]



print(solution(	["I 16", "I -5643", "D -1", "D 1", "D 1", "I 123", "D -1"]))

File: 힙/프로그래머스_이중우선순위큐.js
function solution(operations) {
    const qq = []
    operations.forEach(operation => {
        let [key, value] = operation.split(' ')
        value = parseInt(value)
        switch (key) {
            case 'I':
                qq.unshift(value)  
                break
                
            case 'D':
                if (qq.length > 0) {
                    value > 0 ? qq.sort((a,b) => a - b) : qq.sort((a,b) => b - a) 
                    qq.pop()
                
                }
         
        }
    })
    var answer = qq.length > 0? [Math.max(...qq), Math.min(...qq)] : [0,0];
    return answer;
}